pkg/check/datastore.go

```diff
@@ -0,0 +1,259 @@
+				}
+			}
+		}
+	}
+	klog.V(2).Infof("CheckStorageClasses checked %d storage classes, %d problems found", len(scs), len(errs))
+	return JoinErrors(errs)
+}
+
+// CheckPVs tests that datastore name in all PVs in the cluster is short enough.
+func CheckPVs(ctx *CheckContext) error {
+	var errs []error
+
+	pvs, err := ctx.KubeClient.ListPVs(ctx.Context)
+	if err != nil {
+		return err
+	}
+	for i := range pvs {
+		pv := &pvs[i]
+		if pv.Spec.VsphereVolume == nil {
+			continue
+		}
+		klog.V(4).Infof("Checking PV %s: %s", pv.Name, pv.Spec.VsphereVolume.VolumePath)
+		err := checkVolumeName(pv.Spec.VsphereVolume.VolumePath)
+		if err != nil {
+			klog.V(2).Infof("CheckPVs: %s: %s", pv.Name, err)
+			errs = append(errs, fmt.Errorf("PersistentVolume %s: %s", pv.Name, err))
+		}
+	}
+	klog.V(2).Infof("CheckPVs: checked %d PVs, %d problems found", len(pvs), len(errs))
+	return JoinErrors(errs)
+}
+
+// CheckDefaultDatastore checks that the default data store name in vSphere config file is short enough.
+func CheckDefaultDatastore(ctx *CheckContext) error {
+	infra, err := ctx.KubeClient.GetInfrastructure(ctx.Context)
+	if err != nil {
+		return err
+	}
+
+	dsName := ctx.VMConfig.Workspace.DefaultDatastore
+	if err := checkDataStore(dsName, infra); err != nil {
+		return fmt.Errorf("defaultDatastore %q in vSphere configuration: %s", dsName, err)
+	}
+	return nil
+}
+
+// checkStoragePolicy lists all compatible datastores and checks their names are short.
+func checkStoragePolicy(ctx *CheckContext, policyName string, infrastructure *configv1.Infrastructure) error {
+	klog.V(4).Infof("Checking storage policy %s", policyName)
+
+	pbm, err := getPolicy(ctx, policyName)
+	if err != nil {
+		return err
+	}
+	if len(pbm) == 0 {
+		return fmt.Errorf("error listing storage policy %s: policy not found", policyName)
+	}
+	if len(pbm) > 1 {
+		return fmt.Errorf("error listing storage policy %s: multiple (%d) policies found", policyName, len(pbm))
+	}
+
+	dataStores, err := getPolicyDatastores(ctx, pbm[0].GetPbmProfile().ProfileId)
+	if err != nil {
+		return err
+	}
+	klog.V(4).Infof("Policy %q is compatible with datastores %v", policyName, dataStores)
+
+	var errs []error
+	for _, dataStore := range dataStores {
+		err := checkDataStore(dataStore, infrastructure)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("storage policy %s: %s", policyName, err))
+		}
+	}
+	return JoinErrors(errs)
+}
+
+// checkStoragePolicy lists all datastores compatible with given policy.
+func getPolicyDatastores(ctx *CheckContext, profileID types.PbmProfileId) ([]string, error) {
+	tctx, cancel := context.WithTimeout(ctx.Context, *Timeout)
+	defer cancel()
+	c, err := pbm.NewClient(tctx, ctx.VMClient)
+	if err != nil {
+		return nil, err
+	}
+
+	// Load all datastores in vSphere
+	kind := []string{"Datastore"}
+	m := view.NewManager(ctx.VMClient)
+
+	tctx, cancel = context.WithTimeout(ctx.Context, *Timeout)
```

Comment:
I was just thinking that we have to kind of handle these timeouts very carefully. The default 10s timeout may not be enough in some cases to query an object status and that does not necessarily mean that cluster is unhealthy.

Reply:
I agree, but what would be a good number? And should there be different timeouts for slow/fast API calls? I really don't know.

Reply:
We are going to need different timeout for different calls and in some cases a timeout may not be fatal. But in other cases (listing tasks) - a timeout may mean we don't have necessary permissions since vsphere does not return "permission denied" errors.

Reply:
We will handle timeout issues in a follow up.